### PR TITLE
Separate calc join pool shares

### PR DESCRIPTION
**Pull request by hieuvubk**
Closes: #1753

## What is the purpose of the change
Refactor code
Create two different methods calcSingleTokenJoinPoolShare and calcMultipleTokenPoolShare called by CalcJoinPoolShares base on tokenIns length

### CODE REVIEW

File: x/gamm/pool-models/balancer/amm.go [hunk 1]
```diff
@@ -349,27 +349,44 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 	// more complex / don't just alter the state.
 	// We should simplify this logic further in the future, using balancer multi-join equations.
 
+	if tokensIn.Len() == 1 {
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/amm.go:352]: Thanks for this PR @hieuvubk 

I think we might want to delegate this decision to the caller of this function. More specifically, have 2 separate exported methods:
- `CalcMultiTokenInJoinPoolShares`
- `CalcSingleTokenInJoinPoolShares`

However, I'm not sure if defining these 2 methods on the pool interface would make sense. Would these methods also work for stable swap?

Let's wait on @ValarDragon 's feedback on this before making any changes, please.

Comment (alexanderbez) [reply x/gamm/pool-models/balancer/amm.go:352]: I think this idea makes sense 👍

Comment (hieuvubk) [reply x/gamm/pool-models/balancer/amm.go:352]: Thanks for the suggestion!
I have a question. In amm pool model, **JoinPool** function calls to **CalcJoinPoolShares** internally, should we separate **JoinPool** to single and multi join or based on tokensIn length to call

Comment (p0mvn) [reply x/gamm/pool-models/balancer/amm.go:352]: Good point. Let's keep `JoinPool` as is for now and make it decide which "calc" method to call

For more context and as per https://github.com/osmosis-labs/osmosis/pull/1750, we are considering splitting the `JoinPool` abstraction to `JoinPool` and `JoinPoolNoSwap`.

If we make single and multi-asset join methods and also adopt #1750, there will be 4 different join pool variations.

We can investigate later to determine the best path forward. For now, to limit the scope, let's only deal with the calc methods, please

File: x/gamm/types/pool.go [hunk 2]
```diff
@@ -59,9 +59,13 @@ type PoolI interface {
...
 
-	// CalcJoinPoolShares returns how many LP shares JoinPool would return on these arguments.
+	// CalcSingleTokenInJoinPoolShares returns how many LP shares JoinPool would return on these arguments( single token join).
 	// This does not mutate the pool, or state.
-	CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error)
+	CalcSingleTokenInJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error)
+
+	// CalcMultiTokenInJoinPoolShares returns how many LP shares JoinPool would return on these arguments( multi tokens join).
+	// This does not mutate the pool, or state.
+	CalcMultiTokenInJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error)
```
Comment (ValarDragon) [inline x/gamm/types/pool.go:68]: hrmm, what if we kept the same `CalcJoinPoolShares` API, but then made the first lines of the function doing the split you wrote

Comment (hieuvubk) [reply x/gamm/types/pool.go:68]: https://github.com/osmosis-labs/osmosis/pull/1854/commits/909568221759a11b14cc69ca25f8fcc4039134bd
should revert to this commit?

Comment (AlpinYukseloglu) [reply x/gamm/types/pool.go:68]: I agree with @ValarDragon on adding two functions for `CalcJoinPoolShares` to split into based on its input, and I think splitting into single-token and multi-token joins would be be appropriate abstractions

I'm personally not convinced on the purpose of removing and replacing `CalcJoinPoolShares` altogether though, as it seems to me like we would just be floating the same logic in it up to `JoinPool`, which is already complex enough as it is

So +1 for keeping `CalcJoinPoolShares` and making it internally route to `CalcSingleTokenInJoinPoolShares` and `CalcMultiTokenInJoinPoolShares` based on the passed in token set

Comment (hieuvubk) [reply x/gamm/types/pool.go:68]: https://github.com/osmosis-labs/osmosis/pull/1854/commits/1be8c4b359c2e395cde657403711424ea3e89629
Can u check this commit

Comment (hieuvubk) [reply x/gamm/types/pool.go:68]: Thanks for the suggestion

**Comment by alexanderbez** [comment]
Hey @hieuvubk! Where are we at with this PR? Do you need any assistance on pushing it through? Or is this blocked or perhaps no longer relevant?

**Comment by hieuvubk** [comment]
> Hey @hieuvubk! Where are we at with this PR? Do you need any assistance on pushing it through? Or is this blocked or perhaps no longer relevant?

Sorry i just missed this pr :(((. Still keep CalcJoinPoolShares and call internal function depend on tokenIns length

**Comment by ValarDragon** [comment]
Closing this in light of ^ 

Thanks so much for the work on this!

**Pull request closed without merging**